Failing tests when sentry-go is outside of GOPATH

When the package is checked out outside of GOPATH, some tests fail:

```
$ go test
--- FAIL: TestContextifyFrames (0.00s)
    integrations_test.go:181:
        got: []string{"}", "", "// NewStacktrace creates a stacktrace using `runtime.Callers`.", "func NewStacktrace() *Stacktrace {", "\tpcs := make([]uintptr, 100)"}
        want: []string{")", "", "// NOTE: if you modify this file, you are also responsible for updating LoC position in Stacktrace tests", "", "func Trace() *Stacktrace {"}
    integrations_test.go:188:
        got: "\tn := runtime.Callers(1, pcs)"
        want: "\treturn NewStacktrace()"
    integrations_test.go:189:
        got: []string{"", "\tif n == 0 {", "\t\treturn nil", "\t}", ""}
        want: []string{"}", "", "func RedPkgErrorsRanger() error {", "\treturn BluePkgErrorsRanger()", "}"}
--- FAIL: TestNewStacktrace (0.00s)
    stacktrace_test.go:32:
        got: 3
        want: 2
--- FAIL: TestStacktraceFrame (0.00s)
    stacktrace_test.go:50:
        got: "stacktrace.go"
        want: "errors_test.go"
    stacktrace_test.go:51:
        got: "/Users/rodolfo/sentry/sentry-go/stacktrace.go"
        want: "/Users/rodolfo/sentry/sentry-go/errors_test.go"
    stacktrace_test.go:52:
        got: "NewStacktrace"
        want: "Trace"
    stacktrace_test.go:53:
        got: 36
        want: 12
FAIL
exit status 1
FAIL    github.com/getsentry/sentry-go  0.232s
```

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -3,6 +3,7 @@ linters:
   disable:
     - megacheck
     - stylecheck
+    - funlen
 run:
   skip-dirs:
     - echo

--- a/.travis.yml
+++ b/.travis.yml
@@ -26,7 +26,6 @@ jobs:
         go test -race
   allow_failures:
     - go: master
-    - name: "Module support outside of GOPATH"
   fast_finish: true
 
 before_install:

--- a/integrations_test.go
+++ b/integrations_test.go
@@ -1,6 +1,7 @@
 package sentry
 
 import (
+	"path/filepath"
 	"regexp"
 	"testing"
 )
@@ -174,9 +175,24 @@ func TestContextifyFrames(t *testing.T) {
 		contextLines: 5,
 	}
 
-	stacktrace := Trace()
+	filename := "errors_test.go"
+	abspath, err := filepath.Abs("errors_test.go")
+	if err != nil {
+		t.Fatal(err)
+	}
 
-	frame := cfi.contextify(stacktrace.Frames)[len(stacktrace.Frames)-1]
+	frames := cfi.contextify([]Frame{{
+		Function: "Trace",
+		Module:   "github.com/getsentry/sentry-go",
+		Filename: filename,
+		AbsPath:  abspath,
+		Lineno:   12,
+		InApp:    true,
+	}})
+	if len(frames) != 1 {
+		t.Fatalf("got %d frames, want 1", len(frames))
+	}
+	frame := frames[0]
 
 	assertEqual(t, frame.PreContext, []string{
 		")",

--- a/stacktrace.go
+++ b/stacktrace.go
@@ -4,18 +4,11 @@ import (
 	"go/build"
 	"path/filepath"
 	"reflect"
-	"regexp"
 	"runtime"
 	"strings"
 )
 
 const unknown string = "unknown"
-
-//nolint: gochecknoglobals
-var (
-	isTestFileRegexp    = regexp.MustCompile(`getsentry/sentry-go/.+_test.go`)
-	isExampleFileRegexp = regexp.MustCompile(`getsentry/sentry-go/example/`)
-)
 
 // The module download is split into two parts: downloading the go.mod and downloading the actual code.
 // If you have dependencies only needed for tests, then they will show up in your go.mod,
@@ -213,11 +206,7 @@ func filterFrames(frames []Frame) []Frame {
 			continue
 		}
 		// sentry internal frames
-		isTestFile := isTestFileRegexp.MatchString(frame.AbsPath)
-		isExampleFile := isExampleFileRegexp.MatchString(frame.AbsPath)
-		if strings.Contains(frame.AbsPath, "github.com/getsentry/sentry-go") &&
-			!isTestFile &&
-			!isExampleFile {
+		if frame.Module == "github.com/getsentry/sentry-go" {
 			continue
 		}
 		filteredFrames = append(filteredFrames, frame)

--- a/stacktrace_external_test.go
+++ b/stacktrace_external_test.go
@@ -1,0 +1,228 @@
+package sentry_test
+
+import (
+	"path/filepath"
+	"testing"
+
+	goErrors "github.com/go-errors/errors"
+	pingcapErrors "github.com/pingcap/errors"
+	pkgErrors "github.com/pkg/errors"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/getsentry/sentry-go"
+)
+
+func f1() *sentry.Stacktrace {
+	return sentry.NewStacktrace()
+}
+
+func f2() *sentry.Stacktrace {
+	return f1()
+}
+func f3() *sentry.Stacktrace {
+	return sentry.NewStacktraceForTest()
+}
+
+func RedPkgErrorsRanger() error {
+	return BluePkgErrorsRanger()
+}
+
+func BluePkgErrorsRanger() error {
+	return pkgErrors.New("this is bad from pkgErrors")
+}
+
+func RedPingcapErrorsRanger() error {
+	return BluePingcapErrorsRanger()
+}
+
+func BluePingcapErrorsRanger() error {
+	return pingcapErrors.New("this is bad from pingcapErrors")
+}
+
+func RedGoErrorsRanger() error {
+	return BlueGoErrorsRanger()
+}
+
+func BlueGoErrorsRanger() error {
+	return goErrors.New("this is bad from goErrors")
+}
+
+//nolint: scopelint // false positive https://github.com/kyoh86/scopelint/issues/4
+func TestNewStacktrace(t *testing.T) {
+	tests := map[string]struct {
+		f    func() *sentry.Stacktrace
+		want *sentry.Stacktrace
+	}{
+		"f1": {f1, &sentry.Stacktrace{
+			Frames: []sentry.Frame{
+				{
+					Function: "f1",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   18,
+					InApp:    true,
+				},
+			},
+		}},
+		"f2": {f2, &sentry.Stacktrace{
+			Frames: []sentry.Frame{
+				{
+					Function: "f2",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   22,
+					InApp:    true,
+				},
+				{
+					Function: "f1",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   18,
+					InApp:    true,
+				},
+			},
+		}},
+		// test that functions in the SDK that call NewStacktrace are not part
+		// of the resulting Stacktrace.
+		"NewStacktraceForTest": {sentry.NewStacktraceForTest, &sentry.Stacktrace{
+			Frames: []sentry.Frame{},
+		}},
+		"f3": {f3, &sentry.Stacktrace{
+			Frames: []sentry.Frame{
+				{
+					Function: "f3",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   25,
+					InApp:    true,
+				},
+			},
+		}},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := tt.f()
+			compareStacktrace(t, got, tt.want)
+		})
+	}
+}
+
+//nolint: scopelint // false positive https://github.com/kyoh86/scopelint/issues/4
+func TestExtractStacktrace(t *testing.T) {
+	tests := map[string]struct {
+		f    func() error
+		want *sentry.Stacktrace
+	}{
+		// https://github.com/pkg/errors
+		"pkg/errors": {RedPkgErrorsRanger, &sentry.Stacktrace{
+			Frames: []sentry.Frame{
+				{
+					Function: "RedPkgErrorsRanger",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   29,
+					InApp:    true,
+				},
+				{
+					Function: "BluePkgErrorsRanger",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   33,
+					InApp:    true,
+				},
+			},
+		}},
+		// https://github.com/pingcap/errors
+		"pingcap/errors": {RedPingcapErrorsRanger, &sentry.Stacktrace{
+			Frames: []sentry.Frame{
+				{
+					Function: "RedPingcapErrorsRanger",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   37,
+					InApp:    true,
+				},
+				{
+					Function: "BluePingcapErrorsRanger",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   41,
+					InApp:    true,
+				},
+			},
+		}},
+		// https://github.com/go-errors/errors
+		"go-errors/errors": {RedGoErrorsRanger, &sentry.Stacktrace{
+			Frames: []sentry.Frame{
+				{
+					Function: "RedGoErrorsRanger",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   45,
+					InApp:    true,
+				},
+				{
+					Function: "BlueGoErrorsRanger",
+					Module:   "github.com/getsentry/sentry-go_test",
+					Filename: "stacktrace_external_test.go",
+					AbsPath:  "ignored",
+					Lineno:   49,
+					InApp:    true,
+				},
+			},
+		}},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			err := tt.f()
+			if err == nil {
+				t.Fatal("got nil error")
+			}
+			got := sentry.ExtractStacktrace(err)
+			compareStacktrace(t, got, tt.want)
+		})
+	}
+}
+
+func compareStacktrace(t *testing.T, got, want *sentry.Stacktrace) {
+	t.Helper()
+
+	if len(got.Frames) == 0 {
+		t.Fatal("got no frames")
+	}
+	// Skip anonymous function passed to t.Run.
+	got.Frames = got.Frames[1:]
+
+	if diff := stacktraceDiff(want, got); diff != "" {
+		t.Fatalf("Stacktrace mismatch (-want +got):\n%s", diff)
+	}
+
+	// Because stacktraceDiff ignores Frame.AbsPath, sanity check that the
+	// values we got are actually absolute paths pointing to this test file.
+	for _, frame := range got.Frames {
+		if !filepath.IsAbs(frame.AbsPath) {
+			t.Errorf("Frame{Function: %q}.AbsPath = %q, want absolute path", frame.Function, frame.AbsPath)
+		}
+		if filepath.Base(frame.AbsPath) != "stacktrace_external_test.go" {
+			t.Errorf(`Frame{Function: %q}.AbsPath = %q, want ".../stacktrace_external_test.go"`, frame.Function, frame.AbsPath)
+		}
+	}
+}
+
+func stacktraceDiff(x, y *sentry.Stacktrace) string {
+	return cmp.Diff(
+		x, y,
+		cmpopts.IgnoreFields(sentry.Frame{}, "AbsPath"),
+	)
+}

--- a/stacktrace_test.go
+++ b/stacktrace_test.go
@@ -1,10 +1,13 @@
 package sentry
 
 import (
-	"path/filepath"
 	"runtime"
 	"testing"
 )
+
+func NewStacktraceForTest() *Stacktrace {
+	return NewStacktrace()
+}
 
 func TestFunctionName(t *testing.T) {
 	for _, test := range []struct {
@@ -26,67 +29,8 @@ func TestFunctionName(t *testing.T) {
 	}
 }
 
-func TestNewStacktrace(t *testing.T) {
-	stacktrace := Trace()
-
-	assertEqual(t, len(stacktrace.Frames), 2)
-	assertEqual(t, stacktrace.Frames[0].Function, "TestNewStacktrace")
-	assertEqual(t, stacktrace.Frames[0].Filename, "stacktrace_test.go")
-	assertEqual(t, stacktrace.Frames[1].Function, "Trace")
-	assertEqual(t, stacktrace.Frames[1].Filename, "errors_test.go")
-}
-
 func BenchmarkNewStacktrace(b *testing.B) {
 	for i := 0; i < b.N; i++ {
 		Trace()
 	}
-}
-
-func TestStacktraceFrame(t *testing.T) {
-	_, callerFile, _, _ := runtime.Caller(0)
-	dir, _ := filepath.Split(callerFile)
-
-	stacktrace := Trace()
-	frame := stacktrace.Frames[len(stacktrace.Frames)-1]
-
-	assertEqual(t, frame.Filename, "errors_test.go")
-	assertEqual(t, frame.AbsPath, filepath.Join(dir, "errors_test.go"))
-	assertEqual(t, frame.Function, "Trace")
-	assertEqual(t, frame.Lineno, 12)
-	assertEqual(t, frame.InApp, true)
-	// Go1.10 reports different Module than Modules aware versions (>=1.11)
-	assertStringContains(t, frame.Module, "sentry-go")
-}
-
-// https://github.com/pkg/errors
-func TestExtractStacktracePkgErrors(t *testing.T) {
-	err := RedPkgErrorsRanger()
-	stacktrace := ExtractStacktrace(err)
-
-	assertEqual(t, len(stacktrace.Frames), 3)
-	assertEqual(t, stacktrace.Frames[0].Function, "TestExtractStacktracePkgErrors")
-	assertEqual(t, stacktrace.Frames[1].Function, "RedPkgErrorsRanger")
-	assertEqual(t, stacktrace.Frames[2].Function, "BluePkgErrorsRanger")
-}
-
-// https://github.com/pingcap/errors
-func TestExtractStacktracePingcapErrors(t *testing.T) {
-	err := RedPingcapErrorsRanger()
-	stacktrace := ExtractStacktrace(err)
-
-	assertEqual(t, len(stacktrace.Frames), 3)
-	assertEqual(t, stacktrace.Frames[0].Function, "TestExtractStacktracePingcapErrors")
-	assertEqual(t, stacktrace.Frames[1].Function, "RedPingcapErrorsRanger")
-	assertEqual(t, stacktrace.Frames[2].Function, "BluePingcapErrorsRanger")
-}
-
-// https://github.com/go-errors/errors
-func TestExtractStacktraceGoErrors(t *testing.T) {
-	err := RedGoErrorsRanger()
-	stacktrace := ExtractStacktrace(err)
-
-	assertEqual(t, len(stacktrace.Frames), 3)
-	assertEqual(t, stacktrace.Frames[0].Function, "TestExtractStacktraceGoErrors")
-	assertEqual(t, stacktrace.Frames[1].Function, "RedGoErrorsRanger")
-	assertEqual(t, stacktrace.Frames[2].Function, "BlueGoErrorsRanger")
 }

